File: protocol-generator/ruby/test/conformance_test.rb
# frozen_string_literal: true

require_relative "test_helper"

class ConformanceTest < Minitest::Test
  def test_ecf_conformance_corpus_byte_identical
    bytes = File.binread(CorpusPaths.conformance_corpus)
    results = EntityCore::Conformance.run(bytes)

    failures = results.select { |r| r.status == :fail }
    passes = results.count { |r| r.status == :pass }

    unless failures.empty?
      puts "\n=== ECF conformance failures (#{failures.size}/#{results.size}) ==="
      failures.each { |r| puts "  #{r.id}: #{r.detail}" }
    end

    assert_empty failures, "#{failures.size} vector(s) failed; #{passes}/#{results.size} passed"
    assert_equal results.size, passes
    puts "\nECF corpus: #{passes}/#{results.size} PASS"
  end
end

File: protocol-generator/elixir/test/conformance_test.exs
defmodule EntityCore.ConformanceTest do
  use ExUnit.Case, async: true

  alias EntityCore.Conformance

  # Corpus path resolves from the project root (mix test CWD). Override with the
  # CORPUS env var to point at a different vendored version.
  @corpus System.get_env("CORPUS") ||
            Path.join(["..", "shared", "test-vectors", "v0.8.0", "conformance-vectors-v1.cbor"])

  test "ECF conformance corpus — byte-identical encode + decode-reject" do
    bytes = File.read!(@corpus)
    results = Conformance.run(bytes)

    failures = for {id, {:fail, detail}} <- results, do: {id, detail}
    passes = Enum.count(results, fn {_id, r} -> r == :pass end)

    if failures != [] do
      IO.puts("\n=== ECF conformance failures (#{length(failures)}/#{length(results)}) ===")
      for {id, detail} <- failures, do: IO.puts("  #{id}: #{inspect(detail, limit: :infinity)}")
    end

    assert failures == [], "#{length(failures)} vector(s) failed; #{passes}/#{length(results)} passed"
    assert passes == length(results)
    IO.puts("\nECF corpus: #{passes}/#{length(results)} PASS")
  end
end
